container/multi-value-slice/multi_value_slice.go

```diff
@@ -0,0 +1,380 @@
+type Value[V any] struct {
+	val V
+	ids []interfaces.Id
+}
+
+// MultiValue defines a collection of Value items.
+type MultiValue[V any] struct {
+	Values []*Value[V]
+}
+
+// Slice is the main component of the multi-value slice data structure. It has two type parameters:
+//   - V comparable - the type of values stored the slice. The constraint is required
+//     because certain operations (e.g. updating, appending) have to compare values against each other.
+//   - O interfaces.Identifiable - the type of objects sharing the slice. The constraint is required
+//     because we need a way to compare objects against each other in order to know which objects
+//     values should be accessed.
+type Slice[V comparable, O interfaces.Identifiable] struct {
+	sharedItems     []V
+	individualItems map[uint64]*MultiValue[V]
+	appendedItems   []*MultiValue[V]
+	cachedLengths   map[interfaces.Id]int
+	lock            sync.RWMutex
+}
+
+// Init initializes the slice with sensible defaults. Input values are assigned to shared items.
+func (s *Slice[V, O]) Init(items []V) {
+	s.sharedItems = items
+	s.individualItems = map[interfaces.Id]*MultiValue[V]{}
+	s.appendedItems = []*MultiValue[V]{}
+	s.cachedLengths = map[interfaces.Id]int{}
+}
+
+// Len returns the number of items for the input object.
+func (s *Slice[V, O]) Len(obj O) int {
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+
+	l, ok := s.cachedLengths[obj.Id()]
+	if !ok {
+		return len(s.sharedItems)
+	}
+	return l
+}
+
+// Copy copies items between the source and destination.
+func (s *Slice[V, O]) Copy(src O, dst O) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	for _, item := range s.individualItems {
+		for _, v := range item.Values {
+			found := compareIds(v.ids, src.Id(), func(_ int) {
+				v.ids = append(v.ids, dst.Id())
+			})
+			if found {
+				break
+			}
+		}
+	}
+
+appendedLoop:
+	for _, item := range s.appendedItems {
+		found := false
+		for _, v := range item.Values {
+			found = compareIds(v.ids, src.Id(), func(_ int) {
+				v.ids = append(v.ids, dst.Id())
+			})
+			if found {
+				break
+			}
+		}
+		if !found {
+			// This is an optimization. If we didn't find an appended item at index i,
+			// then all larger indices don't have an appended item for the object either.
+			break appendedLoop
+		}
+	}
+
+	srcLen, ok := s.cachedLengths[src.Id()]
+	if ok {
+		s.cachedLengths[dst.Id()] = srcLen
+	}
+}
+
+// Value returns all items for the input object.
+func (s *Slice[V, O]) Value(obj O) []V {
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+
+	result := make([]V, len(s.sharedItems))
+	for i, item := range s.sharedItems {
+		ind, ok := s.individualItems[uint64(i)]
+		if !ok {
+			result[i] = item
+		} else {
+			found := false
+			for _, v := range ind.Values {
+				found = compareIds(v.ids, obj.Id(), func(_ int) {
+					result[i] = v.val
+				})
+				if found {
+					break
+				}
+			}
+			if !found {
+				result[i] = item
+			}
+		}
+	}
+
+	for _, item := range s.appendedItems {
```

Comment:
There's a lot of looping over items, values, doing compare and break if found. This could be abstracted into its own helper function that could be reused several times in this package